src/analyses/loop_termination.ml

```diff
@@ -0,0 +1,116 @@
+(** Work in progress *)
+
+open Analyses
+open GoblintCil
+open TerminationPreprocessing
+
+exception PreProcessing of string
+
+(*
+let loop_heads () =
+  let module FileCfg =
+  struct
+    let file = !Cilfacade.current_file
+    module Cfg = (val !MyCFG.current_cfg)
+  end in
+  let module WitnessInvariant = WitnessUtil.Invariant (FileCfg) in
+  WitnessInvariant.loop_heads (* TODO: Unused *)
+*)
+
+(** Contains all loop counter variables (varinfo) and maps them to their corresponding loop statement. *)
+let loop_counters : stmt VarToStmt.t ref = ref VarToStmt.empty
+
+(** Contains the locations of the upjumping gotos *)
+let upjumping_gotos : location list ref = ref []
+
+(** Indicates the place in the code, right after a loop is exited. *)
+let loop_exit : varinfo ref = ref (makeVarinfo false "-error" Cil.intType)
+
+let is_loop_counter_var (x : varinfo) =
+  VarToStmt.mem x !loop_counters
+
+let is_loop_exit_indicator (x : varinfo) =
+  x = !loop_exit
+
+let no_upjumping_gotos () =
+  upjumping_gotos.contents = []
+
+(** Checks whether a variable can be bounded *)
+let check_bounded ctx varinfo =
+  let open IntDomain.IntDomTuple in
+  let exp = Lval (Var varinfo, NoOffset) in
+  match ctx.ask (EvalInt exp) with
+  | `Top -> false
+  | `Lifted v -> not (is_top_of (ikind v) v)
+  | `Bot -> raise (PreProcessing "Loop variable is Bot")
+
+module UnitV =
+struct
+  include Printable.Unit
+  include StdV
+end
+
+(** We want to record termination information of loops and use the loop
+ * statements for that. We use this lifting because we need to have a
+ * lattice. *)
+module Statements = Lattice.Flat (CilType.Stmt) (Printable.DefaultNames) (* TODO: Use Basetype.CilStmt instead? *)
+
+(** The termination analysis considering loops and gotos *)
+module Spec : Analyses.MCPSpec =
+struct
+
+  (** Provides some default implementations *)
+  include Analyses.IdentitySpec
+
+  let name () = "termination"
+
+  module D = Lattice.Unit
+  module C = D
+  module V = UnitV
+  module G = MapDomain.MapBot (Statements) (BoolDomain.MustBool)
+
+  let startstate _ = D.bot ()
+  let exitstate = startstate
+
+  let assign ctx (lval : lval) (rval : exp) =
+    (* Detect assignment to loop counter variable *)
```

Comment:
Is it necessary to always do this or does it suffice to do this once at the end? During iteration, precision may be lost and only restored later during narrowing.

Reply:
Originally, we did this check directly after the loop. The reason we moved it into the loop was that the analysis might be unsound in cases like your "pathological loop".

Reply:
Originally, we placed this check right after the loop. The reason we moved it inside the loop was that the analysis might be unsound in cases like your "pathological loop" otherwise.

Reply:
We discussed & resolved this at the meeting today.